File: Path Planning/Pathfinder/GridGraph.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Drawing;

namespace Pathfinder
{

    public class GridGraph
    {
        public int GridSize { get; private set; }
        public HashSet<Point> Nodes { get; private set; }

        public GridGraph(int gridSize)
        {
            Nodes = new HashSet<Point>();
            GridSize = gridSize;
            SetBlock(new Point(0, 0), new Point(gridSize - 1, gridSize - 1), true);
        }

        public void SetBlock(Point p1, Point p2, bool add = false)
        {
            //TODO: check that p1, p2 are in right order!
            for (int x = p1.X; x <= p2.X; x++)
            {
                for (int y = p1.Y; y <= p2.Y; y++)
                {
                    Point p = new Point(x, y);
                    if (add)
                    {
                        Nodes.Add(p);
                    }
                    else
                    {
                        Nodes.Remove(p);
                    }
                }
            }
        }


        private bool NodeExists(Point p)
        {
            return Nodes.Contains(p);
        }

        private bool EdgeExists(Point p1, Point p2)
        {
            return NodeExists(p1) && NodeExists(p2) && !p1.Equals(p2) && (PointsAreAdjacentOrthogonally(p1, p2) || PointsAreAdjacentDiagonally(p1, p2));
        }

        private bool PointsAreAdjacentDiagonally(Point p1, Point p2)
        {
            return Math.Abs(p1.X - p2.X) == 1 && Math.Abs(p1.Y - p2.Y) == 1;
        }

        private bool PointsAreAdjacentOrthogonally(Point p1, Point p2)
        {
            return Math.Abs(p1.X - p2.X) == 1 && (p1.Y == p2.Y) || Math.Abs(p1.Y - p2.Y) == 1 && (p1.X == p2.X);
        }

        public double? FindEdge(Point p1, Point p2)
        {
            if (EdgeExists(p1, p2))
            {
                if (PointsAreAdjacentDiagonally(p1, p2))
                {
                    return Math.Sqrt(2);
                }
                else
                {
                    return 1;
                }

            }
            else
            {
                return null;
            }
        }

        //Returns a list of nodes that can be reached from the specified node
        public List<Point> Neighbours(Point p)
        {
            var list = new List<Point>() {
                new Point(p.X, p.Y-1),
                new Point(p.X+1, p.Y-1),
                new Point(p.X+1, p.Y),
                new Point(p.X+1, p.Y+1),
                new Point(p.X, p.Y+1),
                new Point(p.X-1, p.Y+1),
                new Point(p.X-1, p.Y),
                new Point(p.X-1, p.Y-1) };
            return list.Where(n => NodeExists(n)).ToList();
        }

        //Returns a list of nodes, from source to destination, representing the shortest path, using Dijkstra's algorithm
        //You may make use of existing methods such a s Neignbours, or FindEdge.
        public List<Point> ShortestPath(Point source, Point destination, Algorithms alg)
        {
            //Initialise the 'table' with three 'columns' - one 'row' per node
            Dictionary<Point, bool> visited = NewDictionaryOfAllPointsReturningFalseValues();
            Dictionary<Point, double> costFromSource = NewDictionaryOfAllPointsReturningDoublesSetToInfinity();
            Dictionary<Point, Point?> via = NewDictionaryOfAllPointsReturningNull();
            //Set start
            var currentNode = source;
            costFromSource[currentNode] = 0;
            //Iterate until shortest path found
            while (currentNode != destination)
            {
                visited[currentNode] = true;
                UpdateCostAndViaOfEachNeighbourIfApplicable(costFromSource, via, currentNode, destination);
                currentNode = NextNodeToVisit(currentNode, visited, costFromSource, destination, alg);
                if (costFromSource[currentNode] == double.PositiveInfinity)
                {
                    throw new Exception("Cannot reach destination -  graph is 'disconnected'");
                }
            }
            return RetraceRoute(destination, source, via);
        }

        public Dictionary<Point, Point?> NewDictionaryOfAllPointsReturningNull()
        {
            var dict = new Dictionary<Point, Point?>();
            foreach (Point p in Nodes)
            {
                dict.Add(p, null);
            }
            return dict;
        }

        public Dictionary<Point, double> NewDictionaryOfAllPointsReturningDoublesSetToInfinity()
        {
            var dict = new Dictionary<Point, double>();
            foreach (Point p in Nodes)
            {
                dict.Add(p, double.PositiveInfinity);
            }
            return dict;
        }

        public Dictionary<Point, bool> NewDictionaryOfAllPointsReturningFalseValues()
        {
            var dict = new Dictionary<Point, bool>();
            foreach (Point p in Nodes)
            {
                dict.Add(p, false);
            }
            return dict;
        }

        public void UpdateCostAndViaOfEachNeighbourIfApplicable(Dictionary<Point, double> costFromSource, Dictionary<Point, Point?> via, Point currentNode, Point destination)
        {
            foreach (var neighbour in Neighbours(currentNode))
            {
                double newCost = costFromSource[currentNode] + FindEdge(currentNode, neighbour).Value;

                if (newCost < costFromSource[neighbour])
                {
                    costFromSource[neighbour] = newCost; //Wrong -  don't update cost to estimate -  update to actual. Use estimate to choose the lowest cost unvisited node only. 
                    via[neighbour] = currentNode;
                }
            }
        }

        public double EstimatedCostToDestination(Point current, Point destination)
        {
            return Math.Sqrt(Math.Pow(current.X - destination.X, 2) + Math.Pow(current.Y - destination.Y, 2));
        }

        public Point NextNodeToVisit(Point currentNode, Dictionary<Point, bool> visited, Dictionary<Point, double> costFromSource, Point destination, Algorithms alg)
        {
            var lowestCostSoFar = double.PositiveInfinity;
            Point lowestCostNode = Nodes.First();
            var possibilities = Nodes.Where(n => !visited[n] && costFromSource[n] < double.PositiveInfinity);
            if (possibilities.Count() ==0)
            {
                throw new Exception("The graph is disconnected -  there are no routes from start to destination.");
            }
            foreach (Point p in possibilities)
            {
                double cost = 0;
                switch (alg)
                {
                    case Algorithms.Dijkstra:
                        cost = costFromSource[p];
                        break;
                    case Algorithms.Optimistic:
                        cost = EstimatedCostToDestination(p, destination);
                        break;
                    case Algorithms.AStar:
                        cost = costFromSource[p] + EstimatedCostToDestination(p, destination);
                        break;
                    default:
                        break;
                }
                if (cost < lowestCostSoFar)
                {
                    lowestCostSoFar = cost;
                    lowestCostNode = p;
                }
            }

            return lowestCostNode;
        }

        public List<Point> RetraceRoute(Point destination, Point source, Dictionary<Point, Point?> via)
        {
            var result = new List<Point>() { destination };
            var currentNode = destination;
            while (currentNode != source)
            {
                var previous = via[currentNode];
                result.Insert(0, previous.Value);
                currentNode = previous.Value;
            }
            return result;
        }

        public double SumOfEdges(List<Point> route)
        {
            double result = 0;
            int step = 0;
            while (step < route.Count - 1)
            {
                result += FindEdge(route[step], route[step + 1]).Value;
                step++;
            }
            return result;
        }
    }
}

File: Path Planning/PathfinderVB/GridGraph.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Drawing

Namespace Pathfinder
    Public Class GridGraph
        Public Property GridSize As Integer
        Public Property Nodes As HashSet(Of Point)

        Public Sub New(ByVal gs As Integer)
            Nodes = New HashSet(Of Point)()
            GridSize = gs
            SetBlock(New Point(0, 0), New Point(GridSize - 1, GridSize - 1), True)
        End Sub

        Public Sub SetBlock(ByVal p1 As Point, ByVal p2 As Point, ByVal Optional add As Boolean = False)
            For x As Integer = p1.X To p2.X

                For y As Integer = p1.Y To p2.Y
                    Dim p As Point = New Point(x, y)

                    If add Then
                        Nodes.Add(p)
                    Else
                        Nodes.Remove(p)
                    End If
                Next
            Next
        End Sub

        Private Function NodeExists(ByVal p As Point) As Boolean
            Return Nodes.Contains(p)
        End Function

        Private Function EdgeExists(ByVal p1 As Point, ByVal p2 As Point) As Boolean
            Return NodeExists(p1) AndAlso NodeExists(p2) AndAlso Not p1.Equals(p2) AndAlso (PointsAreAdjacentOrthogonally(p1, p2) OrElse PointsAreAdjacentDiagonally(p1, p2))
        End Function

        Private Function PointsAreAdjacentDiagonally(ByVal p1 As Point, ByVal p2 As Point) As Boolean
            Return Math.Abs(p1.X - p2.X) = 1 AndAlso Math.Abs(p1.Y - p2.Y) = 1
        End Function

        Private Function PointsAreAdjacentOrthogonally(ByVal p1 As Point, ByVal p2 As Point) As Boolean
            Return Math.Abs(p1.X - p2.X) = 1 AndAlso (p1.Y = p2.Y) OrElse Math.Abs(p1.Y - p2.Y) = 1 AndAlso (p1.X = p2.X)
        End Function

        Public Function FindEdge(ByVal p1 As Point, ByVal p2 As Point) As Double?
            If EdgeExists(p1, p2) Then

                If PointsAreAdjacentDiagonally(p1, p2) Then
                    Return Math.Sqrt(2)
                Else
                    Return 1
                End If
            Else
                Return Nothing
            End If
        End Function

        Public Function Neighbours(ByVal p As Point) As List(Of Point)
            Dim list = New List(Of Point)() From {
                New Point(p.X, p.Y - 1),
                New Point(p.X + 1, p.Y - 1),
                New Point(p.X + 1, p.Y),
                New Point(p.X + 1, p.Y + 1),
                New Point(p.X, p.Y + 1),
                New Point(p.X - 1, p.Y + 1),
                New Point(p.X - 1, p.Y),
                New Point(p.X - 1, p.Y - 1)
            }
            Return list.Where(Function(n) NodeExists(n)).ToList()
        End Function

        Public Function ShortestPath(ByVal source As Point, ByVal destination As Point, ByVal alg As Algorithms) As List(Of Point)
            Dim visited As Dictionary(Of Point, Boolean) = NewDictionaryOfAllPointsReturningFalseValues()
            Dim costFromSource As Dictionary(Of Point, Double) = NewDictionaryOfAllPointsReturningDoublesSetToInfinity()
            Dim via As Dictionary(Of Point, Point?) = NewDictionaryOfAllPointsReturningNull()
            Dim currentNode = source
            costFromSource(currentNode) = 0

            While currentNode <> destination
                visited(currentNode) = True
                UpdateCostAndViaOfEachNeighbourIfApplicable(costFromSource, via, currentNode, destination)
                currentNode = NextNodeToVisit(currentNode, visited, costFromSource, destination, alg)

                If costFromSource(currentNode) = Double.PositiveInfinity Then
                    Throw New Exception("Cannot reach destination -  graph is 'disconnected'")
                End If
            End While

            Return RetraceRoute(destination, source, via)
        End Function

        Public Function NewDictionaryOfAllPointsReturningNull() As Dictionary(Of Point, Point?)
            Dim dict = New Dictionary(Of Point, Point?)()

            For Each p As Point In Nodes
                dict.Add(p, Nothing)
            Next

            Return dict
        End Function

        Public Function NewDictionaryOfAllPointsReturningDoublesSetToInfinity() As Dictionary(Of Point, Double)
            Dim dict = New Dictionary(Of Point, Double)()

            For Each p As Point In Nodes
                dict.Add(p, Double.PositiveInfinity)
            Next

            Return dict
        End Function

        Public Function NewDictionaryOfAllPointsReturningFalseValues() As Dictionary(Of Point, Boolean)
            Dim dict = New Dictionary(Of Point, Boolean)()

            For Each p As Point In Nodes
                dict.Add(p, False)
            Next

            Return dict
        End Function

        Public Sub UpdateCostAndViaOfEachNeighbourIfApplicable(ByVal costFromSource As Dictionary(Of Point, Double), ByVal via As Dictionary(Of Point, Point?), ByVal currentNode As Point, ByVal destination As Point)
            For Each neighbour In Neighbours(currentNode)
                Dim newCost As Double = costFromSource(currentNode) + FindEdge(currentNode, neighbour).Value

                If newCost < costFromSource(neighbour) Then
                    costFromSource(neighbour) = newCost
                    via(neighbour) = currentNode
                End If
            Next
        End Sub

        Public Function EstimatedCostToDestination(ByVal current As Point, ByVal destination As Point) As Double
            Return Math.Sqrt(Math.Pow(current.X - destination.X, 2) + Math.Pow(current.Y - destination.Y, 2))
        End Function

        Public Function NextNodeToVisit(ByVal currentNode As Point, ByVal visited As Dictionary(Of Point, Boolean), ByVal costFromSource As Dictionary(Of Point, Double), ByVal destination As Point, ByVal alg As Algorithms) As Point
            Dim lowestCostSoFar = Double.PositiveInfinity
            Dim lowestCostNode As Point = Nodes.First()
            Dim possibilities = Nodes.Where(Function(n) Not visited(n) AndAlso costFromSource(n) < Double.PositiveInfinity)

            If possibilities.Count() = 0 Then
                Throw New Exception("The graph is disconnected -  there are no routes from start to destination.")
            End If

            For Each p As Point In possibilities
                Dim cost As Double = 0

                Select Case alg
                    Case Algorithms.Dijkstra
                        cost = costFromSource(p)
                    Case Algorithms.Optimistic
                        cost = EstimatedCostToDestination(p, destination)
                    Case Algorithms.AStar
                        cost = costFromSource(p) + EstimatedCostToDestination(p, destination)
                    Case Else
                End Select

                If cost < lowestCostSoFar Then
                    lowestCostSoFar = cost
                    lowestCostNode = p
                End If
            Next

            Return lowestCostNode
        End Function

        Public Function RetraceRoute(ByVal destination As Point, ByVal source As Point, ByVal via As Dictionary(Of Point, Point?)) As List(Of Point)
            Dim result = New List(Of Point)() From {
                destination
            }
            Dim currentNode = destination

            While currentNode <> source
                Dim previous = via(currentNode)
                result.Insert(0, previous.Value)
                currentNode = previous.Value
            End While

            Return result
        End Function

        Public Function SumOfEdges(ByVal route As List(Of Point)) As Double
            Dim result As Double = 0
            Dim [step] As Integer = 0

            While [step] < route.Count - 1
                result += FindEdge(route([step]), route([step] + 1)).Value
                [step] += 1
            End While

            Return result
        End Function
    End Class
End Namespace
